File: packages/engine/src/normalize.ts
import { DEFAULTS, PARAM_META } from "./constants";
import type { SimParams, Warning } from "./types";

export interface NormalizedParams {
  params: SimParams;
  warnings: Warning[];
}

export function normalizeParams(input: Partial<SimParams>): NormalizedParams {
  const params: SimParams = { ...DEFAULTS };
  const warnings: Warning[] = [];
  const keys = Object.keys(PARAM_META) as Array<keyof SimParams>;

  const assign = <K extends keyof SimParams>(key: K, value: SimParams[K]): void => {
    params[key] = value;
  };

  for (const key of keys) {
    if (!(key in input)) {
      continue;
    }

    const meta = PARAM_META[key];
    const defaultValue = DEFAULTS[key];
    const raw = input[key];

    if (typeof defaultValue === "number") {
      let next = typeof raw === "number" && Number.isFinite(raw) ? raw : defaultValue;
      const lower = "min" in meta ? meta.min : undefined;
      const upper = "max" in meta ? meta.max : undefined;
      let clamped = false;

      if (typeof lower === "number" && next < lower) {
        next = lower;
        clamped = true;
      }
      if (typeof upper === "number" && next > upper) {
        next = upper;
        clamped = true;
      }
      if (typeof raw !== "number" || !Number.isFinite(raw)) {
        clamped = true;
      }

      assign(key, next as SimParams[typeof key]);
      if (clamped) {
        warnings.push({
          id: "param-clamped",
          severity: "info",
          module: "params",
          message: "Parameter was clamped to its configured bounds.",
          value: typeof raw === "number" && Number.isFinite(raw) ? raw : defaultValue,
          limit: next
        });
      }
      continue;
    }

    if (key === "site") {
      if (raw === "equatorial" || raw === "polar") {
        assign("site", raw);
      } else {
        assign("site", DEFAULTS.site);
        warnings.push({
          id: "param-clamped",
          severity: "info",
          module: "params",
          message: "Parameter was clamped to its configured bounds.",
          value: 0,
          limit: 0
        });
      }
      continue;
    }

    if (key === "storageStream") {
      const options = new Set(["auto", "lox", "water-ice", "liquid-water", "lh2", "lch4", "co2-feed", "custom"]);
      if (typeof raw === "string" && options.has(raw)) {
        assign("storageStream", raw as SimParams["storageStream"]);
      } else {
        assign("storageStream", DEFAULTS.storageStream);
        warnings.push({ id: "param-clamped", severity: "info", module: "params", message: "Parameter was reset to a supported option.", value: 0, limit: 0 });
      }
      continue;
    }

    if (key === "polarProfileMode") {
      if (raw === "scalar" || raw === "profile") {
        assign("polarProfileMode", raw);
      } else {
        assign("polarProfileMode", DEFAULTS.polarProfileMode);
        warnings.push({ id: "param-clamped", severity: "info", module: "params", message: "Parameter was reset to a supported option.", value: 0, limit: 0 });
      }
      continue;
    }

    if (key === "polarProfileData") {
      if (typeof raw === "string" && raw.length <= 100_000) {
        assign("polarProfileData", raw);
      } else {
        assign("polarProfileData", DEFAULTS.polarProfileData);
        warnings.push({ id: "param-clamped", severity: "info", module: "params", message: "Imported profile data exceeded the supported text boundary.", value: 0, limit: 100_000 });
      }
      continue;
    }

    if (key === "cryoControlMode") {
      const options = new Set(["zero-boiloff", "passive", "capacity-limited"]);
      if (typeof raw === "string" && options.has(raw)) {
        assign("cryoControlMode", raw as SimParams["cryoControlMode"]);
      } else {
        assign("cryoControlMode", DEFAULTS.cryoControlMode);
        warnings.push({ id: "param-clamped", severity: "info", module: "params", message: "Parameter was reset to a supported option.", value: 0, limit: 0 });
      }
      continue;
    }

    if (typeof defaultValue === "boolean") {
      if (typeof raw === "boolean") {
        assign(key, raw as SimParams[typeof key]);
      } else {
        assign(key, defaultValue as SimParams[typeof key]);
        warnings.push({
          id: "param-clamped",
          severity: "info",
          module: "params",
          message: "Parameter was clamped to its configured bounds.",
          value: 0,
          limit: 0
        });
      }
    }
  }

  return { params, warnings };
}

File: python/selene_isru/normalize.py
from __future__ import annotations

import math
from typing import Any

from .constants import DEFAULTS, PARAM_META


def normalize_params(input_params: dict[str, Any] | None = None) -> tuple[dict[str, Any], list[dict[str, Any]]]:
    source = {} if input_params is None else input_params
    params = dict(DEFAULTS)
    warnings: list[dict[str, Any]] = []

    for key, meta in PARAM_META.items():
        if key not in source:
            continue

        raw = source[key]
        default_value = DEFAULTS[key]

        if isinstance(default_value, (int, float)) and not isinstance(default_value, bool):
            next_value = float(raw) if isinstance(raw, (int, float)) and not isinstance(raw, bool) and math.isfinite(raw) else float(default_value)
            lower = meta.get("min")
            upper = meta.get("max")
            clamped = not (isinstance(raw, (int, float)) and not isinstance(raw, bool) and math.isfinite(raw))

            if isinstance(lower, (int, float)) and next_value < float(lower):
                next_value = float(lower)
                clamped = True
            if isinstance(upper, (int, float)) and next_value > float(upper):
                next_value = float(upper)
                clamped = True

            params[key] = next_value
            if clamped:
                warnings.append(
                    {
                        "id": "param-clamped",
                        "severity": "info",
                        "module": "params",
                        "message": "Parameter was clamped to its configured bounds.",
                        "value": float(raw) if isinstance(raw, (int, float)) and not isinstance(raw, bool) and math.isfinite(raw) else float(default_value),
                        "limit": next_value,
                    }
                )
            continue

        if key == "site":
            if raw in ("equatorial", "polar"):
                params["site"] = raw
            else:
                params["site"] = DEFAULTS["site"]
                warnings.append(
                    {
                        "id": "param-clamped",
                        "severity": "info",
                        "module": "params",
                        "message": "Parameter was clamped to its configured bounds.",
                        "value": 0,
                        "limit": 0,
                    }
                )
            continue

        string_options = {
            "storageStream": {"auto", "lox", "water-ice", "liquid-water", "lh2", "lch4", "co2-feed", "custom"},
            "cryoControlMode": {"zero-boiloff", "passive", "capacity-limited"},
            "polarProfileMode": {"scalar", "profile"},
        }
        if key in string_options:
            if isinstance(raw, str) and raw in string_options[key]:
                params[key] = raw
            else:
                params[key] = default_value
                warnings.append({"id": "param-clamped", "severity": "info", "module": "params", "message": "Parameter was reset to a supported option.", "value": 0, "limit": 0})
            continue

        if key == "polarProfileData":
            if isinstance(raw, str) and len(raw) <= 100_000:
                params[key] = raw
            else:
                params[key] = default_value
                warnings.append({"id": "param-clamped", "severity": "info", "module": "params", "message": "Imported profile data exceeded the supported text boundary.", "value": 0, "limit": 100_000})
            continue

        if isinstance(default_value, bool):
            if isinstance(raw, bool):
                params[key] = raw
            else:
                params[key] = default_value
                warnings.append(
                    {
                        "id": "param-clamped",
                        "severity": "info",
                        "module": "params",
                        "message": "Parameter was clamped to its configured bounds.",
                        "value": 0,
                        "limit": 0,
                    }
                )

    return params, warnings
